app/data_sources/bookstack.py

```diff
@@ -0,0 +1,179 @@
+import concurrent.futures
+import logging
+from datetime import datetime
+from typing import List, Dict
+
+from data_source_api.basic_document import BasicDocument, DocumentType
+from data_source_api.base_data_source import BaseDataSource
+from data_source_api.exception import InvalidDataSourceConfig
+from indexing_queue import IndexingQueue
+from parsers.html import html_to_text
+from pydantic import BaseModel
+from requests import Session
+from requests.auth import AuthBase
+from urllib.parse import urljoin
+
+
+class BookStackAuth(AuthBase):
+    def __init__(self, token_id, token_secret, header_key="Authorization"):
+        self.header_key = header_key
+        self.token_id = token_id
+        self.token_secret = token_secret
+
+    def __call__(self, r):
+        r.headers[self.header_key] = f"Token {self.token_id}:{self.token_secret}"
+        return r
+
+
+class BookStack(Session):
+    def __init__(self, url: str, token_id: str, token_secret: str, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.base_url = url
+        self.auth = BookStackAuth(token_id, token_secret)
+
+    def request(self, method, url_path, *args, **kwargs):
+        url = urljoin(self.base_url, url_path)
+        return super().request(method, url, *args, **kwargs)
+
+    def get_books(self, count: int = 500, offset: int = 0):
+        r = self.get("/api/books", params={"count": count, "offset": offset, "sort": "+updated_at"},
+                     headers={"Content-Type": "application/json"})
+        data = r.json().get("data")
+        records = len(data)
+        if records >= count:
```

Comment:
`count=500` and you get only `130` books,  you still wanna add them right?

Reply:
If you get less than `count` you don't need to recursively fetch again with an increased `offset` ?

Reply:
oh my bad, I see what you did there.
Can you write it in a non-recursive way tho? a `while has_more` or somth might be more readable. - thanks

Reply:
Done in [2140f9f](https://github.com/GerevAI/gerev/pull/21/commits/2140f9feeaa06d8e8ed694c4489e689439b271ef)